### ISSUE TITLE
django channels periodic disconnection due to error reporting:JSONDecodeError

### ISSUE THREAD

**Issue by Davelyb** [django/channels#1339]
When I start Django channels locally using the python management.py runserver command, and then the front end connects via websocket javascript, the following error is reported every few seconds:
WebSocket DISCONNECT /ws/ [10.203.42.114:51011]
WebSocket HANDSHAKING /ws/ [10.203.42.114:51100]
WebSocket CONNECT /ws/ [10.203.42.114:51100]
Exception inside application: Expecting value: line 1 column 1 (char 0)
  File "/Users/liuyabo/.local/share/virtualenvs/bs_channelsbackend_graphql-57bnZmLE/lib/python3.7/site-packages/channels/sessions.py", line 179, in __call__
    return await self.inner(receive, self.send)
  File "/Users/liuyabo/.local/share/virtualenvs/bs_channelsbackend_graphql-57bnZmLE/lib/python3.7/site-packages/channels/middleware.py", line 41, in coroutine_call
    await inner_instance(receive, send)
  File "/Users/liuyabo/.local/share/virtualenvs/bs_channelsbackend_graphql-57bnZmLE/lib/python3.7/site-packages/channels/consumer.py", line 59, in __call__
    [receive, self.channel_receive], self.dispatch
  File "/Users/liuyabo/.local/share/virtualenvs/bs_channelsbackend_graphql-57bnZmLE/lib/python3.7/site-packages/channels/utils.py", line 52, in await_many_dispatch
    await dispatch(result)
  File "/Users/liuyabo/.local/share/virtualenvs/bs_channelsbackend_graphql-57bnZmLE/lib/python3.7/site-packages/channels/consumer.py", line 73, in dispatch
    await handler(message)
  File "/Users/liuyabo/.local/share/virtualenvs/bs_channelsbackend_graphql-57bnZmLE/lib/python3.7/site-packages/channels/generic/websocket.py", line 196, in websocket_receive
    await self.receive(text_data=message["text"])
  File "/Users/liuyabo/.local/share/virtualenvs/bs_channelsbackend_graphql-57bnZmLE/lib/python3.7/site-packages/channels/generic/websocket.py", line 259, in receive
    await self.receive_json(await self.decode_json(text_data), **kwargs)
  File "/Users/liuyabo/.local/share/virtualenvs/bs_channelsbackend_graphql-57bnZmLE/lib/python3.7/site-packages/channels/generic/websocket.py", line 277, in decode_json
    return json.loads(text_data)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
  Expecting value: line 1 column 1 (char 0)
WebSocket DISCONNECT /ws/ [10.203.42.114:51100]
WebSocket HANDSHAKING /ws/ [10.203.42.114:51193]
WebSocket CONNECT /ws/ [10.203.42.114:51193]

has anyone ever encountered this problem?

Environment:
OS: mac pro 10.13

Python: 3.7
Main python module:
channels==2.1.6
channels-redis==2.3.2
daphne==2.2.3
Django==2.1.4

yarn: 1.13.0

**Comment by carltongibson** [comment]
Can you inspect the `message` that's being sent? 

```
 await self.receive(text_data=message["text"])
```

It doesn't look like it's valid. (But hard to say from the info here.) 

This doesn't look like a channels issue per se. Happy to consider if it you can narrow it down to a failing example.

**Comment by Davelyb** [comment]
> Can you inspect the `message` that's being sent?
> 
> ```
>  await self.receive(text_data=message["text"])
> ```
> 
> It doesn't look like it's valid. (But hard to say from the info here.)
> 
> This doesn't look like a channels issue per se. Happy to consider if it you can narrow it down to a failing example.

Thank you for your answer. Finally, I solved the problem.  
Reason: By tracing the source code of channels, I found that the front end would send a heartbeat message all the time, and the heartbeat message format was just a string. But the channel's receive function decodes all the information it receives  by calling the json. loads () function. this place(json. loads ('heartbeat')) causes the above error. So let the front end modify the heartbeat format as a dictionary object, and then I can process the heartbeat information in a custom function.
